_posts/2017-03-17-getting-rest-right.md

```diff
@@ -0,0 +1,120 @@
+
+You may have noticed from the above URL schema that I have gone for singular URL nouns across the board. There is still some debate about whether singular or plural nouns are preferable, with plural seeming to win out but with most admitting it's down to individual preference.
+
+Most agree that you should stick to one or the other across the whole of your API, to keep the amount of learning to a minimum.
+
+Verbs should not show up in your API URLs - if possible. _"But what about `/search`?"_ I hear you cry.
+
+Well if search is specific to an account then maybe having something like `/article?q=search+term` may be more appropriate to limit the results of the article list. Alternatively if you do have global search endpoint, then it's probably appropriate to go for the `/search` endpoint.
+
+## Versioning
+
+Unless your API is for internal use only, and will only be used by your team, you should version your API from the first public release.
+
+There are a couple of different ways you can go about this.
+
+1. In URL versioning (http://api.example.com/v1/article/43)
+   - Probably the most popular, but unfortunately semi-breaks the "URLs represent a resource" principle
+2. Custom header (X-API-Version: v1)
+   - Developers have to learn a new header just to use your API.
+3. Accept header (Accept: application/vnd.article-v1+json)
+   - Developers can't explore your API just using a browser.
+
+There's no easy answer, but I usually go for the compromise of in URL versioning for easy exploration of the API in combination with...
+
+## HATEOAS
+
+Standing for the easy to remember "Hypermedia As The Engine Of Application State", it's a fancy way of saying - return URLs to further resources in your API.
+
+If you have a request to /v1/article, it should return something similar to the following:
+
+```JSON
+GET /v1/article HTTP/1.1
+Host: api.example.com
+Accept: application/json
+
+[
+  {
+    "id": 42,
+    "name": "HG2G - Don't Panic",
+    "url": "/v1/article/42"
+  },
+  {
+    "id": 256,
+    "name": "Two to the Eight",
+    "url": "/v1/article/256"
+  }
+]
+```
+
+The url property of the article object allows the consumer to know how view the detail of the article object, without having to look at your documentation (you did ~~write~~ generate documentation right?). If you're using Postman, you can just click on the link and it will take you straight to the next resource!
+
+## SSL Everywhere
+
+With the invention of Let's Encrypt, there is no longer any excuse not to secure your API endpoint.
+
+If you've got HTTPS enabled, and you receive a request to HTTP - just throw an error such as `400 Bad Request {"error": "HTTPS Required"}` rather than redirecting to HTTPS.
+
+## Errors and HTTP status codes
+
+There are more [HTTP status codes](https://en.wikipedia.org/wiki/List_of_HTTP_status_codes) than you can shake a stick at. Also all good libraries should be able to read these, so you may as well utilise them.
```

Comment:
Different libraries interpret the codes differently though, e.g. throw an exception on non-200 status codes which can be ... surprising.